# Exploring Data

### Introduction

Now that we know how to work with a dataframe and select individual columns, it's time for us to see if we can begin to understand our data.

### Exploring a DataFrame

So let's get a better sense of the data in our FEMA dataset.  The first thing we'll do is load up the data, and then perhaps look at the columns available.

In [1]:
import pandas as pd
url = 'houston_claims.csv'
claims_df = pd.read_csv(url, index_col = 0)
claims_df[:3]

reportedCity                dateOfLoss elevatedBuildingIndicator floodZone  \
0      HOUSTON  2017-08-27T00:00:00.000Z                     False         X   
1      HOUSTON  2008-09-12T00:00:00.000Z                     False         X   
2      HOUSTON  2004-06-29T00:00:00.000Z                     False         X   

   latitude  longitude  lowestFloodElevation  amountPaidOnBuildingClaim  \
0      29.7      -95.5                   NaN                  195857.43   
1      29.5      -95.1                   NaN                       0.00   
2      29.8      -95.6                   NaN                    1420.89   

   amountPaidOnContentsClaim                yearofLoss  reportedZipcode  \
0                        0.0  2017-01-01T00:00:00.000Z            77096   
1                        0.0  2008-01-01T00:00:00.000Z            77058   
2                        0.0  2004-01-01T00:00:00.000Z            77042   

                         id  
0  5e398d6774cbd479fc898dea  
1  5e398d6774cbd479fc898dfc  
2  5e398d6774cbd479fc898e4b

### Viewing Sample Statistics

In pandas, we can view sample statistics either on a column or across a dataframe.  For example, we can look at the average `amountPaidOnBuildingClaim` in the dataset.

In [3]:
claims_df.amountPaidOnBuildingClaim.mean()

66239.67425949442

> So we can see that the average amount paid on a claim is `66239.67425949442`.

Or we can look at the mean across our numeric columns.

In [4]:
claims_df.mean()

elevatedBuildingIndicator        0.046052
latitude                        29.779968
longitude                      -95.448202
lowestFloodElevation            61.329824
amountPaidOnBuildingClaim    66239.674259
amountPaidOnContentsClaim    20042.823729
reportedZipcode              77040.538700
dtype: float64

> Notice that because we asked for a numerical statistic, pandas only calculated the average for the columns with numerical values.

Now if we want to get an overview of the data in each of the columns, we can do so with the describe method.

In [7]:
claims_df.describe()

latitude     longitude  lowestFloodElevation  \
count  19943.000000  19943.000000           4939.000000   
mean      29.779968    -95.448202             61.329824   
std        0.312829      0.487151             26.201233   
min       29.500000   -149.800000             -1.000000   
25%       29.700000    -95.500000             49.000000   
50%       29.800000    -95.500000             59.000000   
75%       29.800000    -95.400000             78.000000   
max       61.600000    -80.200000            252.000000   

       amountPaidOnBuildingClaim  amountPaidOnContentsClaim  reportedZipcode  
count               1.966900e+04               16442.000000     20000.000000  
mean                6.623967e+04               20042.823729     77040.538700  
std                 9.022818e+04               37712.538422      1090.962214  
min                 0.000000e+00                   0.000000     15342.000000  
25%                 2.250000e+03                   0.000000     77033.000000  
50%                 3.282040e+04                2311.275000     77062.000000  
75%                 9.234882e+04               25600.000000     77084.000000  
max                 1.436755e+06              500000.000000     99694.000000

> So we can see the mean and standard deviation, as well as the mode (at the 50% level).

As we can see, this shows us the `mean` (that is, the average), and the standard deviation (which we'll describe later), as well as the range and percentiles.

* Working with categorical values

Now the above functions are good for summarizing numeric data.  Now let's talk about categorical data.  For example, with `reportedZipcode`, we may want to see what are some of the most frequent zipcodes.  We can do so, with the value_counts method.

In [8]:
claims_df['reportedZipcode'].value_counts()[:10]

77096    1881
77079    1039
77025     855
77024     849
77084     693
77089     682
77074     672
77035     536
77088     461
77070     455
Name: reportedZipcode, dtype: int64

> So we can see that the most frequent zipcode is 77096, occurring 1881 times.

Let's try this again with `floodZone`.

> We first select the `floodZone` series, and then we call the `value_counts` method on that series.

In [14]:
claims_df['floodZone'].value_counts()[:10]

X      10276
AE      8708
C        363
A        273
AOB       94
AO        75
B         45
A04       40
A05       30
A10       20
Name: floodZone, dtype: int64

> Note that this method is only available on a series, not a dataframe.

### Summary

In [ ]:
In this lesson